test/e2e/run_networking_test.go

```diff
@@ -543,56 +545,58 @@ EXPOSE 2004-2005/tcp`, ALPINE)
-			session.WaitWithDefaultTimeout()
-			ncListener.WaitWithDefaultTimeout()
-
-			Expect(session).Should(ExitCleanly())
-			Expect(ncListener).Should(Exit(0))
-			Expect(ncListener.ErrorToString()).To(ContainSubstring("Connection from 127.0.0.1"))
-		} else {
-			session := podmanTest.Podman([]string{"run", "--network", networkConfiguration, "-dt", ALPINE, "nc", "-w", "2", "10.0.2.2", port})
-			session.WaitWithDefaultTimeout()
-			Expect(session).To(ExitWithError(125, "outbound_addr not supported"))
+	for _, local := range []bool{true, false} {
+		local := local
+		testName := "HostIP"
+		if local {
+			testName = "127.0.0.1"
 		}
-	})
-
-	It("podman run network bind to HostIP", func() {
-		// Determine our likeliest outgoing IP address
-		conn, err := net.Dial("udp", "8.8.8.8:80")
-		Expect(err).ToNot(HaveOccurred())
+		It(fmt.Sprintf("podman run network slirp4netns bind to %s", testName), func() {
+			ip := "127.0.0.1"
+			if !local {
+				// Determine our likeliest outgoing IP address
+				conn, err := net.Dial("udp", "8.8.8.8:80")
+				Expect(err).ToNot(HaveOccurred())
 
-		defer conn.Close()
-		ip := conn.LocalAddr().(*net.UDPAddr).IP
-		port := strconv.Itoa(GetPort())
+				defer conn.Close()
+				ip = conn.LocalAddr().(*net.UDPAddr).IP.String()
+			}
+			port := strconv.Itoa(GetPort())
 
-		slirp4netnsHelp := SystemExec("slirp4netns", []string{"--help"})
-		Expect(slirp4netnsHelp).Should(ExitCleanly())
-		networkConfiguration := fmt.Sprintf("slirp4netns:outbound_addr=%s,allow_host_loopback=true", ip.String())
-
-		if strings.Contains(slirp4netnsHelp.OutputToString(), "outbound-addr") {
-			ncListener := StartSystemExec("nc", []string{"-v", "-n", "-l", "-p", port})
-			session := podmanTest.Podman([]string{"run", "--network", networkConfiguration, ALPINE, "nc", "-w", "2", "10.0.2.2", port})
-			session.Wait(30)
-			ncListener.Wait(30)
+			networkConfiguration := fmt.Sprintf("slirp4netns:outbound_addr=%s,allow_host_loopback=true", ip)
 
+			listener, err := net.Listen("tcp", ":"+port)
+			Expect(err).ToNot(HaveOccurred())
+			defer listener.Close()
+
+			msg := RandomString(10)
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			// now use a new goroutine to start accepting connection in the background and make the checks there
+			go func() {
+				defer GinkgoRecover()
+				defer wg.Done()
+				conn, err := listener.Accept()
+				Expect(err).ToNot(HaveOccurred())
+				defer conn.Close()
+				addr := conn.RemoteAddr()
+				// addr will be in the form ip:port, we don't care about the port as it is random
+				Expect(addr.String()).To(ContainSubstring(ip+":"), "remote address")
+				gotBytes, err := io.ReadAll(conn)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(string(gotBytes)).To(Equal(msg))
+			}()
+
+			session := podmanTest.Podman([]string{"run", "--network", networkConfiguration, ALPINE, "sh", "-c", "echo -n " + msg + " | nc -w 2 10.0.2.2 " + port})
```

Comment:
Now that I look at this closer... could it be that `-w 2` is too low a timeout for slow CI hosts? Is there a reason for such a low value?

Reply:
Maybe? I am not sure how nc calculates this time-out, in theory with the current version I would expect the connect to work first try as we bind before starting the container

It seems to be documented as `Set a fixed timeout for connection attempts` but if it also applies to actually sending the data then maybe. I guess there is no harm bumping this up to 30s or something like this.